Common/src/grid_movement/CVolumetricMovement.cpp

```diff
@@ -293,6 +296,141 @@ void CVolumetricMovement::ComputeDeforming_Element_Volume(CGeometry *geometry, s
+          }
+        }
+
+        /*--- Calculate cross product of edge vectors ---*/
+        crossProduct = edgeVector_1[1]*edgeVector_2[0] - edgeVector_1[0]*edgeVector_2[1];
+
+        if (crossProduct < minCrossProduct) minCrossProduct = crossProduct;
+        if (crossProduct > maxCrossProduct) maxCrossProduct = crossProduct;
+      }
+      
+      /*--- Element is nonconvex if cross product of at least one set of adjacent edges is negative ---*/
+      if (minCrossProduct < 0 && maxCrossProduct > 0){
+        nNonconvexElements++;
+      }
+    }
+  } else {
+
+    /*--- 3D elements ---*/
+    unsigned short iNode, iFace, nFaceNodes;
+    unsigned long face_point_i, face_point_j, face_point_k;
+    su2double *Coords_i, *Coords_j, *Coords_k;
+    su2double *ev1, *ev2;
+    su2double crossProduct[nDim], crossProductLength, minCrossProductLength, maxCrossProductLength;
+
+    for (iElem = 0; iElem < geometry->GetnElem(); iElem++){
+      for (iFace = 0; iFace < geometry->elem[iElem]->GetnFaces(); iFace++){
+        nFaceNodes = geometry->elem[iElem]->GetnNodesFace(iFace);
+
+        minCrossProductLength = 1.e6; maxCrossProductLength = -1.e6;
+        for (iNode = 0; iNode < nFaceNodes; iNode++) {
+          crossProductLength = 0.0;
+          
+          /*--- Get coords of node face_point_i and its adjacent nodes in the face ---*/
+          face_point_i = geometry->elem[iElem]->GetNode(geometry->elem[iElem]->GetFaces(iFace, iNode));
+
+          if (iNode == 0) {
+            face_point_j = geometry->elem[iElem]->GetNode(geometry->elem[iElem]->GetFaces(iFace, nFaceNodes-1));
+            face_point_k = geometry->elem[iElem]->GetNode(geometry->elem[iElem]->GetFaces(iFace, iNode+1));
+          } else if (iNode == nFaceNodes-1) {
+            face_point_j = geometry->elem[iElem]->GetNode(geometry->elem[iElem]->GetFaces(iFace, iNode-1));
+            face_point_k = geometry->elem[iElem]->GetNode(geometry->elem[iElem]->GetFaces(iFace, 0));
+          }
+          
+          Coords_i = geometry->nodes->GetCoord(face_point_i);
+          Coords_j = geometry->nodes->GetCoord(face_point_j);
+          Coords_k = geometry->nodes->GetCoord(face_point_k);
+
+          /*--- Get edge vectors from point k to i and point j to i ---*/
+          for (iDim = 0; iDim < nDim; iDim++) {
+            edgeVector_1[iDim] = Coords_k[iDim] - Coords_i[iDim];
+            edgeVector_2[iDim] = Coords_j[iDim] - Coords_i[iDim];
+          }
+
+          /*--- Calculate cross product of edge vectors and its length---*/
+          ev1 = edgeVector_1;
+          ev2 = edgeVector_2;
+ 
+          crossProduct[0] = ev1[1]*ev2[2] - ev1[2]*ev2[1];
+          crossProduct[1] = ev1[2]*ev2[0] - ev1[0]*ev2[2];
+          crossProduct[2] = ev1[0]*ev2[1] - ev1[1]*ev2[0];
+ 
+          for (iDim = 0; iDim < nDim; iDim++) {
+            crossProductLength += crossProduct[iDim]*crossProduct[iDim];
+          }
+          
+          crossProductLength = sqrt(crossProductLength);  
```

Comment:
There is a geometry toolbox that implements all these, please use it.

Reply:
Done

Reply:
Thanks! There a few more things that the toolbox could do, like computing distance between 2 points, but we can do that latter.